### PR TITLE
Creating terra-breakpoints package

### PR DESCRIPTION
**Pull request by tbiethman**
### Summary
This PR creates the terra-breakpoints package. Included are:

* Newly redefined breakpoint ranges
* Utility functions to translate width values into breakpoint ranges
* ActiveBreakpointProvider, a component that provides the active breakpoint value to children through context
* SASS mixins for media queries using these breakpoint values.

### Deployment Link
https://terra-core-deployed-pr-2039.herokuapp.com/#/components/terra-breakpoints/breakpoints/about

[CONTRIBUTORS.md]: ../blob/master/CONTRIBUTORS.md
[License]: ../blob/master/LICENSE

Fixes: #2018

### CODE REVIEW

File: packages/terra-breakpoints/.npmignore [hunk 1]
```diff
@@ -0,0 +1,5 @@
+*.log
+node_modules
+target
+reports
+tests/**/__snapshots__/
```
Comment (emilyrohrbough) [inline packages/terra-breakpoints/.npmignore]: might be jumping the guy, but should we include @StephenEsser's new tar file changes?

Comment (tbiethman) [reply packages/terra-breakpoints/.npmignore]: Not of we merge this first :P

Comment (StephenEsser) [reply packages/terra-breakpoints/.npmignore]: giphy 1

File: packages/terra-breakpoints/docs/README.md [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+- Install with [npmjs](https://www.npmjs.com):
+  - `npm install terra-breakpoints`
+
+## Usage
+
+### Breakpoints 
+
+An object containing the named breakpoint values is the default export of the `terra-breakpoints` package. These named breakpoints are defined as minimum values.
+
+|Breakpoint Name|Minimum Value|
```
Comment (bjankord) [inline packages/terra-breakpoints/docs/README.md]: I feel like we should expand this table to show the media query and description.

```
|Breakpoint Name|Minimum Value|Media Query|Description|
|---|---|---|---|
|`tiny`|0px|`@media screen and (min-width: 0px)`| Active from viewport width 0px and up |
|`small`|544px|`@media screen and (min-width: 544px)`|Active from viewport width 544px and up|
|`medium`|768px|`@media screen and (min-width: 768px)`|Active from viewport width 768px and up|
|`large`|992px|`@media screen and (min-width: 992px)`|Active from viewport width 992px and up|
|`huge`|1216px|`@media screen and (min-width: 1216px)`|Active from viewport width 1216 and up|
|`enormous`|1440px|`@media screen and (min-width: 1440px)`|Active from viewport width 1440px and up|
```

File: packages/terra-breakpoints/docs/README.md [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+|---|---|
+|`tiny`|0px|
+|`small`|544px|
+|`medium`|768px|
+|`large`|992px|
+|`huge`|1216px|
+|`enormous`|1440px|
+
+Helper functions are also provided to to help determine what breakpoint is active for a given value.
```
Comment (bjankord) [inline packages/terra-breakpoints/docs/README.md]: nit: Change `Helper functions are also provided to to help` to `Utility functions are also provided to help`

File: packages/terra-breakpoints/docs/README.md [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+    <ActiveBreakpointConsumer1 />
+    <hr />
+    <ActiveBreakpointConsumer2 />
+  </ActiveBreakpointProvider>
+);
+```
+
+### Media Queries
+
+`terra-breakpoints` includes `_media-queries.scss`, which contains a set of SASS mixins that define media queries for the supported breakpoints.
```
Comment (bjankord) [inline packages/terra-breakpoints/docs/README.md]: minor nit: `Sass` instead of `SASS`. http://sassnotsass.com/

File: packages/terra-breakpoints/src/terra-dev-site/doc/example/MixinsExample.jsx [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+      }
+    }
+  }
+*/
+
+const cx = classNames.bind(styles);
+
+const MixinsExample = () => (
+  <div>
+    <p>This example uses SASS mixins to determine the active breakpoint.</p>
```
Comment (bjankord) [inline packages/terra-breakpoints/src/terra-dev-site/doc/example/MixinsExample.jsx]: minor nit: `Sass` instead of `SASS`. http://sassnotsass.com/

**Comment by mjhenkes** [comment]
Do we want to remove the sass file from responsive element to encourage people to use this package?
This would be a breaking change.

**Comment by tbiethman** [comment]
@mjhenkes @bjankord We should definitely remove it, but we also need to uplift the ResponsiveElement to adhere to the new breakpoints (remove default, add enormous, bump everything else down a level). That's going to be aa much larger change, and I'm not sure we want to do that on the eve of the other major version releases. Given the scope of those changes, they should be done is a separate PR regardless.

**Comment by bjankord** [comment]
@tbiethman Agreed, let's handle responsive element changes in a separate PR. I think we can roll up the breaking change of removing the sass file from the responsive element with other breaking changes we are planning in the near future.

File: packages/terra-breakpoints/tests/jest/breakpoints.test.jsx [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+describe('breakpointIsActiveForSize', () => {
+  it('should return false for unknown breakpoint', () => {
+    expect(breakpointIsActiveForSize('test', 100)).toBe(false);
+  });
+
+  it('should return false for negative size', () => {
+    expect(breakpointIsActiveForSize('tiny', -100)).toBe(false);
+  });
+
+  it('should accurately determine default size matching', () => {
```
Comment (JakeLaCombe) [inline packages/terra-breakpoints/tests/jest/breakpoints.test.jsx]: Take this with a grain of salt, and it's a minor suggestion. For edge case testing, perhaps we can test a point in between the breakpoint values if others feel that adds any value (272 for example)

Comment (tbiethman) [reply packages/terra-breakpoints/tests/jest/breakpoints.test.jsx]: I went back and forth on that; as usual, laziness prevailed. I can add some more expectations here though.

File: packages/terra-breakpoints/.npmignore (hunk 1 above)
Comment (emilyrohrbough) [reply packages/terra-breakpoints/.npmignore]: He merged first 😹

Comment (tbiethman) [reply packages/terra-breakpoints/.npmignore]: http://www.nooooooooooooooo.com/
